File: src/lib/siie-import.ts
import type { OrdemSection } from '@/types/ordem-item'

export const EXPECTED_HEADERS = [
  'agrupamento', 'nin', 'nome', 'datanascimento', 'dataadmissao', 'Sexo',
  'Situacao', 'Categoria', 'morada', 'localidade', 'telefone', 'CP 1', 'CP 2',
  'codigopostal', 'telemovel', 'email', 'pai', 'Telefone Pai', 'Email Pai',
  'mae', 'Telefone Mae', 'Email Mae', 'Enc Educ', 'Enc Educ Telefone',
  'Enc Educ Email', 'nif', 'cc',
] as const

export function categoryToSection(categoria: string | null | undefined): OrdemSection | null {
  if (!categoria) return null
  const last = categoria.trim().slice(-1).toUpperCase()
  switch (last) {
    case 'L': return 'ALCATEIA'
    case 'E': return 'EXPEDICAO'
    case 'P': return 'COMUNIDADE'
    case 'C': return 'CLA'
    default: return null // D, AD, ND, DH and anything unknown
  }
}

export function splitName(full: string): { firstName: string; lastName: string } {
  const cleaned = full.replace(/\s+/g, ' ').trim()
  if (!cleaned) return { firstName: '', lastName: '' }
  const parts = cleaned.split(' ')
  if (parts.length === 1) return { firstName: parts[0], lastName: '' }
  return { firstName: parts[0], lastName: parts.slice(1).join(' ') }
}

/**
 * Parses values like:
 *  - JS Date instance (when xlsx is read with cellDates:true)
 *  - "DD/MM/YYYY" string
 *  - "YYYY-MM-DD" ISO string
 *  - Excel date serial number
 */
export function parseExcelDate(value: unknown): Date | null {
  if (value instanceof Date && !isNaN(value.getTime())) return value
  if (typeof value === 'number' && Number.isFinite(value)) {
    // Excel serial: days since 1899-12-30 (sheetjs default)
    const ms = Math.round((value - 25569) * 86400 * 1000)
    const d = new Date(ms)
    return isNaN(d.getTime()) ? null : d
  }
  if (typeof value === 'string') {
    const s = value.trim()
    if (!s) return null
    // DD/MM/YYYY
    const m = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/)
    if (m) {
      const [, dd, mm, yyyy] = m
      const d = new Date(Date.UTC(parseInt(yyyy, 10), parseInt(mm, 10) - 1, parseInt(dd, 10)))
      return isNaN(d.getTime()) ? null : d
    }
    const d = new Date(s)
    return isNaN(d.getTime()) ? null : d
  }
  return null
}

export function combinePostalCode(cp1: unknown, cp2: unknown, locality: unknown): string | null {
  const code = `${String(cp1 ?? '').trim()}-${String(cp2 ?? '').trim().padStart(3, '0')}`.trim()
  const hasCode = code !== '-' && code !== '-000'
  const loc = typeof locality === 'string' ? locality.trim() : ''
  if (!hasCode && !loc) return null
  if (hasCode && loc) return `${code} ${loc}`
  return hasCode ? code : loc
}

export type ImportRow = {
  agrupamento: string | null
  nin: string | null
  nome: string
  datanascimento: unknown
  dataadmissao: unknown
  Sexo: string | null
  Situacao: string | null
  Categoria: string | null
  morada: string | null
  localidade: string | null
  telefone: string | null
  'CP 1': unknown
  'CP 2': unknown
  codigopostal: string | null
  telemovel: string | null
  email: string | null
  pai: string | null
  'Telefone Pai': string | null
  'Email Pai': string | null
  mae: string | null
  'Telefone Mae': string | null
  'Email Mae': string | null
  'Enc Educ': string | null
  'Enc Educ Telefone': string | null
  'Enc Educ Email': string | null
  nif: string | null
  cc: string | null
}

export type ScoutImportPayload = {
  firstName: string
  lastName: string
  numeroAssociado: string | null
  dateOfBirth: Date
  joinedAt: Date | null
  section: OrdemSection | null
  active: boolean
  sexo: string | null
  cc: string | null
  nif: string | null
  email: string | null
  telefone: string | null
  telemovel: string | null
  morada: string | null
  localidade: string | null
  codigoPostal: string | null
  paiNome: string | null
  paiTelefone: string | null
  paiEmail: string | null
  maeNome: string | null
  maeTelefone: string | null
  maeEmail: string | null
  encarregadoNome: string | null
  encarregadoTelefone: string | null
  encarregadoEmail: string | null
}

function cleanStr(v: unknown): string | null {
  if (typeof v !== 'string') {
    if (v == null || v === '') return null
    return String(v).trim() || null
  }
  const t = v.trim()
  return t || null
}

export type MappedRow =
  | { ok: true; nin: string; payload: ScoutImportPayload; categoria: string | null }
  | { ok: false; row: number; nome: string; error: string }

export function mapRow(row: ImportRow, rowIndex: number): MappedRow {
  const nome = typeof row.nome === 'string' ? row.nome.trim() : ''
  if (!nome) return { ok: false, row: rowIndex, nome: '', error: 'Sem nome' }

  const nin = cleanStr(row.nin)
  if (!nin) return { ok: false, row: rowIndex, nome, error: 'Sem NIN — necessário para upsert' }

  const dateOfBirth = parseExcelDate(row.datanascimento)
  if (!dateOfBirth) return { ok: false, row: rowIndex, nome, error: 'Data de nascimento inválida' }

  const joinedAt = parseExcelDate(row.dataadmissao)
  const { firstName, lastName } = splitName(nome)
  const section = categoryToSection(row.Categoria)
  const active = (cleanStr(row.Situacao) ?? '').toUpperCase() === 'A'

  return {
    ok: true,
    nin,
    categoria: cleanStr(row.Categoria),
    payload: {
      firstName,
      lastName,
      numeroAssociado: nin,
      dateOfBirth,
      joinedAt,
      section,
      active,
      sexo: cleanStr(row.Sexo),
      cc: cleanStr(row.cc),
      nif: cleanStr(row.nif),
      email: cleanStr(row.email),
      telefone: cleanStr(row.telefone),
      telemovel: cleanStr(row.telemovel),
      morada: cleanStr(row.morada),
      localidade: cleanStr(row.localidade),
      codigoPostal: combinePostalCode(row['CP 1'], row['CP 2'], row.codigopostal),
      paiNome: cleanStr(row.pai),
      paiTelefone: cleanStr(row['Telefone Pai']),
      paiEmail: cleanStr(row['Email Pai']),
      maeNome: cleanStr(row.mae),
      maeTelefone: cleanStr(row['Telefone Mae']),
      maeEmail: cleanStr(row['Email Mae']),
      encarregadoNome: cleanStr(row['Enc Educ']),
      encarregadoTelefone: cleanStr(row['Enc Educ Telefone']),
      encarregadoEmail: cleanStr(row['Enc Educ Email']),
    },
  }
}

File: api/app/core/siie_import.py
"""SIIE `export.xlsx` row mapping. Port of `@qtscout/core/siie-import`.

The router reads the workbook (openpyxl) into per-row dicts keyed by the SIIE
header names, then `map_row` validates/normalises each into a ScoutImportPayload.
"""

import math
import re
from dataclasses import dataclass
from datetime import date, datetime
from datetime import timedelta as _td
from typing import Any

from app.models.enums import OrdemSection

EXPECTED_HEADERS = (
    "agrupamento", "nin", "nome", "datanascimento", "dataadmissao", "Sexo",
    "Situacao", "Categoria", "morada", "localidade", "telefone", "CP 1", "CP 2",
    "codigopostal", "telemovel", "email", "pai", "Telefone Pai", "Email Pai",
    "mae", "Telefone Mae", "Email Mae", "Enc Educ", "Enc Educ Telefone",
    "Enc Educ Email", "nif", "cc",
)  # fmt: skip


def category_to_section(categoria: str | None) -> OrdemSection | None:
    if not categoria:
        return None
    last = categoria.strip()[-1:].upper()
    return {
        "L": OrdemSection.ALCATEIA,
        "E": OrdemSection.EXPEDICAO,
        "P": OrdemSection.COMUNIDADE,
        "C": OrdemSection.CLA,
    }.get(last)  # D, AD, ND, DH and anything unknown -> None


def split_name(full: str) -> tuple[str, str]:
    cleaned = re.sub(r"\s+", " ", full).strip()
    if not cleaned:
        return "", ""
    parts = cleaned.split(" ")
    if len(parts) == 1:
        return parts[0], ""
    return parts[0], " ".join(parts[1:])


def parse_excel_date(value: Any) -> datetime | None:
    """Accept a datetime/date, an Excel serial number, or a date string."""
    if isinstance(value, datetime):
        return value
    if isinstance(value, date):
        return datetime(value.year, value.month, value.day)
    if isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value):
        # Excel serial: days since 1899-12-30 (sheetjs default).
        return datetime(1970, 1, 1) + _td(seconds=round((value - 25569) * 86400))
    if isinstance(value, str):
        s = value.strip()
        if not s:
            return None
        m = re.match(r"^(\d{1,2})/(\d{1,2})/(\d{4})$", s)
        if m:
            dd, mm, yyyy = (int(g) for g in m.groups())
            try:
                return datetime(yyyy, mm, dd)
            except ValueError:
                return None
        try:
            return datetime.fromisoformat(s.replace("Z", "+00:00")).replace(tzinfo=None)
        except ValueError:
            return None
    return None


def combine_postal_code(cp1: Any, cp2: Any, locality: Any) -> str | None:
    code = f"{str(cp1 or '').strip()}-{str(cp2 or '').strip().rjust(3, '0')}".strip()
    has_code = code not in ("-", "-000")
    loc = locality.strip() if isinstance(locality, str) else ""
    if not has_code and not loc:
        return None
    if has_code and loc:
        return f"{code} {loc}"
    return code if has_code else loc


def clean_str(value: Any) -> str | None:
    if not isinstance(value, str):
        if value is None or value == "":
            return None
        return str(value).strip() or None
    return value.strip() or None


@dataclass
class ScoutImportPayload:
    first_name: str
    last_name: str
    numero_associado: str | None
    date_of_birth: datetime
    joined_at: datetime | None
    section: OrdemSection | None
    active: bool
    sexo: str | None
    cc: str | None
    nif: str | None
    email: str | None
    telefone: str | None
    telemovel: str | None
    morada: str | None
    localidade: str | None
    codigo_postal: str | None
    pai_nome: str | None
    pai_telefone: str | None
    pai_email: str | None
    mae_nome: str | None
    mae_telefone: str | None
    mae_email: str | None
    encarregado_nome: str | None
    encarregado_telefone: str | None
    encarregado_email: str | None


@dataclass
class MapResult:
    ok: bool
    row: int
    nome: str
    nin: str | None = None
    payload: ScoutImportPayload | None = None
    error: str | None = None


def map_row(row: dict[str, Any], row_index: int) -> MapResult:
    nome_raw = row.get("nome")
    nome = nome_raw.strip() if isinstance(nome_raw, str) else ""
    if not nome:
        return MapResult(ok=False, row=row_index, nome="", error="Sem nome")

    nin = clean_str(row.get("nin"))
    if not nin:
        return MapResult(
            ok=False, row=row_index, nome=nome, error="Sem NIN — necessário para upsert"
        )

    date_of_birth = parse_excel_date(row.get("datanascimento"))
    if not date_of_birth:
        return MapResult(ok=False, row=row_index, nome=nome, error="Data de nascimento inválida")

    first_name, last_name = split_name(nome)
    return MapResult(
        ok=True,
        row=row_index,
        nome=nome,
        nin=nin,
        payload=ScoutImportPayload(
            first_name=first_name,
            last_name=last_name,
            numero_associado=nin,
            date_of_birth=date_of_birth,
            joined_at=parse_excel_date(row.get("dataadmissao")),
            section=category_to_section(row.get("Categoria")),
            active=(clean_str(row.get("Situacao")) or "").upper() == "A",
            sexo=clean_str(row.get("Sexo")),
            cc=clean_str(row.get("cc")),
            nif=clean_str(row.get("nif")),
            email=clean_str(row.get("email")),
            telefone=clean_str(row.get("telefone")),
            telemovel=clean_str(row.get("telemovel")),
            morada=clean_str(row.get("morada")),
            localidade=clean_str(row.get("localidade")),
            codigo_postal=combine_postal_code(
                row.get("CP 1"), row.get("CP 2"), row.get("codigopostal")
            ),
            pai_nome=clean_str(row.get("pai")),
            pai_telefone=clean_str(row.get("Telefone Pai")),
            pai_email=clean_str(row.get("Email Pai")),
            mae_nome=clean_str(row.get("mae")),
            mae_telefone=clean_str(row.get("Telefone Mae")),
            mae_email=clean_str(row.get("Email Mae")),
            encarregado_nome=clean_str(row.get("Enc Educ")),
            encarregado_telefone=clean_str(row.get("Enc Educ Telefone")),
            encarregado_email=clean_str(row.get("Enc Educ Email")),
        ),
    )
